parse_weather

ma-quest opened the pull request:
Вариант парсинга погоды
close #122

ma-quest committed:
parse_weather

010721

aerubanov commented:
Привет! А можешь дать мне права коммитить в твою ветку? Чтобы я мог прямо сюда внести правки. Нужно поставить галочку " Allow edits from maintainers" (https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

aerubanov committed:
move script

aerubanov commented:
Я сделал папку отдельную для скрипта и перенес его туда. Смотри, теперь нужно сделать две вещи:
- Записывать результат работы скрипта в файл, скажем в `DATA/raw/forecast/`. Формат можно сделать csv, можно использовать для этого пандас. Создать DataFrame и его записывать в csv. Только предварительно в него добавить столбец с datetime, когда данные получены.
- Нужно сделать так, чтобы загрузка данных осуществлялась через определенный промежуток времени, например, в начале каждого часа, или с интервалом обновления прогноза на сайте (если он больше). Мне кажется удобным для этого использовать модуль [schedule](https://github.com/dbader/schedule).

ma-quest committed:
Update parse_weather.py

добавлен блок с Schedule

aerubanov commented on weather_loader/parse_weather.py:
```diff
@@ -0,0 +1,122 @@
+    frame = pd.DataFrame(columns=['Num','Now','Date-Time','Weather','Temp','Presure','Wind_speed','Wind_dir','Humidity'])
+    if __name__ == '__main__':
+        res = parse_weather(url)
+        for i in range(len(res['date'])):
```
Можно обойтись без этого цикла, а создавать датафрейм сразу из словаря. См. https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.from_dict.html

aerubanov commented on weather_loader/parse_weather.py:
```diff
@@ -0,0 +1,122 @@
+            new_line = {'Num':i, 'Now':datetime.datetime.utcnow(), 'Date-Time':res['date'][i], 'Weather':res['prec'][i], 'Temp':res['temp'][i],'Presure':res['pressure'][i],'Wind_speed':res['wind_speed'][i], 'Wind_dir':res['wind_dir'][i], 'Humidity':res['humidity'][i]}
+            frame = frame.append(new_line, ignore_index=True)
+            #print('frame:', frame)
+    frame.to_csv('parsed_weather.csv', sep=',', header=True, index=False) #запись DataFrame в файл
```
Мне кажется, что в таком виде это будет просто затирать существующий файл каждый раз, когда запускается, а не дописывать в него. Нужно проверить, там вроде должен быть аргумент, с помощью которого это поведение можно настраивать.

ma-quest replied:
да.   нужно , чтобы каждый час автоматом дописывалось по 1 новой строке ?
при первом запуске все 24 часа, а потом каждый час по новой строке ?

aerubanov replied:
Не, там каждый час тоже 24 строки будет. Прогноз ведь снова на сутки будет. Кстати нужно поискать инфу насколько часто прогноз на сайте обновляется, а то может просто нет смысла каждый час его скачивать.

ma-quest replied:
вроде бы там каждый новый час появляется прогноз на следующий 24ый час (видимый период удлиняется)

каждый новый час нужно, чтобы 24 строки следующие появлялись ? они перекрываться будут с предыдущими 23 (повторяться)

или, имеешь в виду , обновлять, когда уже существующий прогноз корректируется/изменяется ?

aerubanov replied:
Да, нужно каждый новый час новые 24 строки. Смотри, сначало был например прогноз на +24 часа вперед, а потом он стал прогнозом уже на +4 часа и там стали другие данные. А мы хотим понять точность прогноза, а она будет скорее всего разная у +24 часа и у +4 часа, поэтому нужны оба значения. Это дает возможность востановить потом историю измений прогноза как-бы.

ma-quest replied:
т.е. у нас будут в файле ; 1ый час - 24 строки, 2ой час - старые 24 строки + 24 новые строки , 3ий - 24+24+24 и тд, при этом получится 21*3 строк на одно и то же время ?

aerubanov replied:
Да. Только ещё нужно для каждой строки записывать время когда она была записана, чтобы потом понять, был ли это прогноз на +1 час, +4 или + 24.

aerubanov commented on weather_loader/parse_weather.py:
```diff
@@ -0,0 +1,122 @@
+    print('done')
+    #frame2 = pd.read_csv('parsed_weather.csv', header=0, sep=',')
+
+#Schedule
```
Посмотри про конструкцию ```if __name__ == "__main__":``` https://stackoverflow.com/questions/419163/what-does-if-name-main-do.

ma-quest replied:
прочитала. нужна ли она вообще в этом скрипте ?

aerubanov replied:
Прямой необходимости нет, мы вроде нигде больше не будем эти функции использовать. Просто лично мне так код читать проще, когда тело самого скрипта внутри ```if __name__ == "__main__":``` и, таким образом, отделено от инициализации переменных в модуле и прочего. Поэтому в глаза бросилось.

ma-quest committed:
Update parse_weather.py

- строка if __name__ == '__main__': перенесена
- при запуске программы создается пустой DataFrame , далее при работе schedule в него дописываются 24 строки
- новые строки дописываются без цикла for
- schedule запускает программу раз в час

aerubanov committed:
some fixes

aerubanov commented:
Смотри, у тебя каждый раз при запуске скрипта в файл с данными писался заголовок, даже если файл уже существует и в нем лежат данные. Я это исправил, и немного форматирование подправил. Теперь возникает такой вопрос: а если в какой-то момент, когда скрипт будет работать на сервере и собирать данные, вместо странички с прогнозом погоды с сайта он получит 404 ошибку например. Что будет?

ma-quest commented:
> Смотри, у тебя каждый раз при запуске скрипта в файл с данными писался заголовок, даже если файл уже существует и в нем лежат данные. Я это исправил, и немного форматирование подправил. Теперь возникает такой вопрос: а если в какой-то момент, когда скрипт будет работать на сервере и собирать данные, вместо странички с прогнозом погоды с сайта он получит 404 ошибку например. Что будет?

aerubanov committed:
add docker files

ma-quest committed:
Update parse_weather.py

добавлена обработка исключений в ф-ию start_parsing

aerubanov committed:
remove SSLError
